EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc

```diff
@@ -0,0 +1,71 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "EventFilter/L1TRawToDigi/plugins/UnpackerFactory.h"
+
+#include "L1Trigger/L1TCalorimeter/interface/CaloTools.h"
+
+#include "L1TObjectCollections.h"
+
+#include "L1TStage2Layer2Constants.h"
+#include "ZDCUnpacker.h"
+
+namespace l1t {
+  namespace stage2 {
+    ZDCUnpacker::ZDCUnpacker() : ZDCSumCopy_(0) {}
+
+    bool ZDCUnpacker::unpack(const Block& block, UnpackerCollections* coll) {
+      using namespace l1t::stage2::layer2;
+
+      LogDebug("L1T") << "Block ID  = " << block.header().getID() << " size = " << block.header().getSize();
+
+      int nBX = int(
+          ceil(block.header().getSize() / demux::nOutputFramePerBX));  // Since there 6 frames per demux output event
+      // expect the first four frames to be the first 4 EtSum objects reported per event (see CMS IN-2013/005)
+
+      // Find the central, first and last BXs
+      int firstBX = -(ceil((double)nBX / 2.) - 1);
+      int lastBX;
+      if (nBX % 2 == 0) {
+        lastBX = ceil((double)nBX / 2.);
+      } else {
+        lastBX = ceil((double)nBX / 2.) - 1;
+      }
+
+      auto res_ = static_cast<L1TObjectCollections*>(coll)->getZDCSums(ZDCSumCopy_);
+      res_->setBXRange(firstBX, lastBX);
+
+      LogDebug("L1T") << "nBX = " << nBX << " first BX = " << firstBX << " lastBX = " << lastBX;
+
+      // Loop over multiple BX and fill EtSums collection
+      for (int bx = firstBX; bx <= lastBX; bx++) {
+        // ZDC -
+        int iFrame = (bx - firstBX) * demux::nOutputFramePerBX;
+
+        uint32_t raw_data = block.payload().at(iFrame +1); // ZDC - info is found on frame 1 of each bx
+        
+        l1t::EtSum zdcm{l1t::EtSum::kZDCM};
+        zdcm.setHwPt(raw_data & 0xFFFF);
+        zdcm.setP4(l1t::CaloTools::p4Demux(&zdcm));

```

Comment:
Just to check here: my understanding is that this function computes the four vector with the implicit understanding that the LSB is 0.5 GeV. Is that the case for the ZDC sums too? (Again this is based on me not finding documentation, so may be a moot point.)

Reply:
it is correct your interpretation but in ZDC we should use the numerical expression and not the conversion in energy as it is going to be wrong, in my understanding. @cfmcginn can comment.

Reply:
Right, I think the uGT emulator always uses the hw values anyway and gets the scales from the menu XML. The physical pT should, however be correct because that's what people usually look at in e.g. the ntuples.

Reply:
what about the ZDC L1 candidate used in the object map ?

Reply:
what about the ZDC L1 candidate used in the object map ?

Reply:
Yes, absolutely. If someone accesses the physical pT value of that object they would get a value of `hw_code*0.5` GeV. I'm not entirely sure from Ivan's answer whether this is correct.